pkg/convert/pprof.go

```diff
@@ -0,0 +1,139 @@
+// Copyright 2022-2023 The Parca Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package convert
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/google/pprof/profile"
+
+	"github.com/parca-dev/parca-agent/pkg/process"
+	"github.com/parca-dev/parca-agent/pkg/profiler"
+)
+
+// TODO(kakkoyun): We can use a pool of pprof here.
+
+type ProfileToPprofConverter struct {
+	mappingCache map[string]*profile.Mapping
+}
+
+func NewProfileToPprofConverter() *ProfileToPprofConverter {
+	return &ProfileToPprofConverter{
+		mappingCache: map[string]*profile.Mapping{},
+	}
+}
+
+// Convert converts several per process Profile to a pprof Profile.
+func (c *ProfileToPprofConverter) Convert(captureTime time.Time, periodNS int64, prs ...*profiler.Profile) (*profile.Profile, error) {
+	prof := &profile.Profile{
+		SampleType: []*profile.ValueType{{
+			Type: "samples",
+			Unit: "count",
+		}},
+		TimeNanos:     captureTime.UnixNano(),
+		DurationNanos: int64(time.Since(captureTime)),
+
+		// Sampling at 100Hz would be every 10 Million nanoseconds.
+		PeriodType: &profile.ValueType{
+			Type: "cpu",
+			Unit: "nanoseconds",
+		},
+		Period: periodNS,
+	}
+	if len(prs) == 0 {
+		return prof, nil
+	}
+
+	// Build Profile from samples, locations and mappings.
```

Comment:
this needs to happen exactly the opposite way, first mappings and functions need to be normalized, then locations then samples, that's the only way we can ensure IDs are correct at the end